amy/dashboard/views.py

```diff
@@ -245,57 +245,135 @@ def training_progress(request):
         context["title"] = str("Submit your Get Involved activity")
         return context
 
     def post(self, request):
         self.object = None
+
         base_training_progress = TrainingProgress(
             trainee=request.user,
             state="n",  # not evaluated yet
             requirement=TrainingRequirement.objects.get(name="Get Involved"),
         )
-        get_involved_form = GetInvolvedForm(
-            data=request.POST, instance=base_training_progress
-        )
+        # select data specifically to prevent user from submitting extra args
+        data = {
+            "involvement_type": request.POST.get("involvement_type"),
+            "date": request.POST.get("date"),
+            "url": request.POST.get("url"),
+            "trainee_notes": request.POST.get("trainee_notes"),
+        }
+        get_involved_form = GetInvolvedForm(data=data, instance=base_training_progress)
 
-        if get_involved_form.is_valid():
+        user_may_create_submission = self.user_may_create_submission(request.user)
+        if get_involved_form.is_valid() and user_may_create_submission:
             return self.form_valid(get_involved_form)
-
         else:
+            if not user_may_create_submission:
+                self.form_invalid_message = (
+                    "You already have an existing submission. "
+                    "You may not create another submission unless your previous "
+                    'submission has the status "asked to repeat."'
+                )
             return self.form_invalid(get_involved_form)
 
 
-class GetInvolvedUpdateView(LoginRequiredMixin, RedirectSupportMixin, AMYUpdateView):
-    # permission_required = TODO
+class GetInvolvedUpdateView(
+    LoginRequiredMixin, PermissionRequiredMixin, RedirectSupportMixin, AMYUpdateView
+):
     model = TrainingProgress
     form_class = GetInvolvedForm
     template_name = "get_involved_form.html"
     success_url = reverse_lazy("training-progress")
     success_message = "Your Get Involved submission was updated successfully."
+    pk_url_kwarg = "pk"
+
+    def has_permission(self):
+        """The user has permission if all the conditions below are met:
+        1. the progress exists
+        2. the progress belongs to the user
+        3. the progress is for the Get Involved step
+        4. the progress has not yet been evaluated
+        """
+        progress_pk = self.kwargs.get(self.pk_url_kwarg)
+        try:
+            progress = TrainingProgress.objects.get(pk=progress_pk)
+            get_involved = TrainingRequirement.objects.get(name="Get Involved")
+            if (
+                progress.trainee == self.request.user
+                and progress.requirement == get_involved
+                and progress.state == "n"
+            ):
+                return True
+            else:
+                return False
+        except TrainingProgress.DoesNotExist:
+            return False
```

Comment:
There's a useful method `self.get_object()` available. It will get correct `TrainingProgress` from the database for you.
If you construct your queryset in a specific way, you'll disable users from getting others' training progresses. This is semantically more correct, as permissions involve e.g. groups and allowance to perform generic action (e.g. edit a training progress), not checking if selected training progress contains specific training requirement.

So I propose to move some of what you have in `has_permission` to queryset. The changes would look like so:

```python
class GetInvolvedUpdateView(
    LoginRequiredMixin, PermissionRequiredMixin, AMYUpdateView
):
    # we could add this anyway, so that in future it's easier to manage permissions
    # - but if users currently don't have these permissions then I'd comment them out
    # or remove `PermissionsRequiredMixin`
    permission_required = ["workshops.view_trainingrequirement", "workshops.change_trainingprogress"]

    # model = TrainingProgress  # not used with queryset

    form_class = GetInvolvedForm
    template_name = "get_involved_form.html"
    success_url = reverse_lazy("training-progress")
    success_message = "Your Get Involved submission was updated successfully."

    def get_queryset(self) -> QuerySet[TrainingProgress]:
        return TrainingProgress.objects.filter(
            trainee=self.request.user,  # you may need to save `self.request = request` in dispatch() or get/post
            requirement__name="Get Involved",
            state="n",
        )

    def get_context_data(self, **kwargs) -> dict:
        context = super().get_context_data(**kwargs)
        context["title"] = str("Update your Get Involved submission")
        return context
```